In [7]:
import pandas as pd
import os

# Set the folder path
folder_path = r"c:\Users\santo\OneDrive\Documents\Coding\Learning\799-02"

# Load the three CSV files into dataframes
predict_conversion_df = pd.read_csv(os.path.join(folder_path, "digital_marketing_campaign_dataset.csv"))
google_ads_df = pd.read_csv(os.path.join(folder_path, "GoogleAds_DataAnalytics_Sales_Uncleaned.csv"))
marketing_performance_df = pd.read_csv(os.path.join(folder_path, "marketing_and_product_performance.csv"))

# Digital Marketing Campaign Dataset - Comprehensive Analysis

## 1. Data Sources and Collection Methods

**Dataset Name:** Predict Conversion in Digital Marketing Dataset

**Source & Nature:** This is a **synthetically generated dataset** created by Rabie El Kharoua for educational purposes. It does not reflect real-world data but provides a realistic simulation of customer behavior and campaign performance metrics.

**Purpose:** Designed for data science and machine learning projects, including:
- Predictive modeling of customer conversion rates
- A/B testing analysis
- Analyzing effectiveness of different marketing channels and campaign types
- Identifying key factors driving customer engagement and conversion

**Data Simulation:** The dataset simulates aggregated metrics from:
- Customer Relationship Management (CRM) systems
- Web analytics platforms
- Email marketing platforms
- Social media analytics tools
- Advertising platforms
- Internal business databases


**Time Period:** No temporal information included (no timestamps/dates)

- Citation Required: Please properly cite the author when using this dataset

**Geographic Scope:** Global (simulated without geographic specificity)- License: CC BY 4.0 (Attribution 4.0 International)

- Author: Rabie El Kharoua

**Data Type:** Customer-level behavioral and campaign performance metrics**Attribution & License:** 


In [8]:
# 2. DATA DESCRIPTION: Summary Statistics, Data Types, and Volume

print("=" * 80)
print("DATASET OVERVIEW")
print("=" * 80)
print(f"\nDataset Dimensions:")
print(f"  - Rows (Samples): {predict_conversion_df.shape[0]:,}")
print(f"  - Columns (Features): {predict_conversion_df.shape[1]}")
print(f"  - Total Data Points: {predict_conversion_df.shape[0] * predict_conversion_df.shape[1]:,}")

print("\n" + "=" * 80)
print("DATA TYPES")
print("=" * 80)
print(predict_conversion_df.dtypes)

print("\n" + "=" * 80)
print("SUMMARY STATISTICS")
print("=" * 80)
print(predict_conversion_df.describe().T)

print("\n" + "=" * 80)
print("FIRST FEW RECORDS")
print("=" * 80)
print(predict_conversion_df.head())


DATASET OVERVIEW

Dataset Dimensions:
  - Rows (Samples): 8,000
  - Columns (Features): 20
  - Total Data Points: 160,000

DATA TYPES
CustomerID               int64
Age                      int64
Gender                  object
Income                   int64
CampaignChannel         object
CampaignType            object
AdSpend                float64
ClickThroughRate       float64
ConversionRate         float64
WebsiteVisits            int64
PagesPerVisit          float64
TimeOnSite             float64
SocialShares             int64
EmailOpens               int64
EmailClicks              int64
PreviousPurchases        int64
LoyaltyPoints            int64
AdvertisingPlatform     object
AdvertisingTool         object
Conversion               int64
dtype: object

SUMMARY STATISTICS
                    count          mean           std           min  \
CustomerID         8000.0  11999.500000   2309.545410   8000.000000   
Age                8000.0     43.625500     14.902785     18.000000   

## 3. DATA DICTIONARY

| Column Name | Data Type | Description | Relevance to Analysis |
|---|---|---|---|
| **CustomerID** | Integer | Unique identifier for each customer | Key for tracking customers and aggregation |
| **Age** | Integer | Customer's age in years | Demographic segmentation, customer profiling |
| **Gender** | String/Categorical | Customer's gender (Male/Female) | Demographic segmentation, campaign personalization |
| **Income** | Numeric | Customer's annual income | Customer segmentation, targeting high-value customers |
| **CampaignChannel** | String/Categorical | Primary marketing channel used (Email, Social, Display, etc.) | **Core metric** - compare performance across channels |
| **CampaignType** | String/Categorical | Type of campaign (Brand Awareness, Conversion, Retention, etc.) | **Core metric** - segment analysis by campaign objective |
| **AdSpend** | Numeric | Amount spent on the advertisement/campaign | Budget efficiency analysis, ROI calculations |
| **ClickThroughRate (CTR)** | Numeric (0-1 scale) | Rate at which customers click on the marketing content | **Key KPI** - campaign engagement metric |
| **ConversionRate** | Numeric (0-1 scale) | Rate at which clicks convert to desired actions (purchases, etc.) | **Key KPI** - campaign effectiveness metric |
| **WebsiteVisits** | Integer | Number of visits to the website | Traffic generation metric |
| **PagesPerVisit** | Numeric | Average number of pages viewed per visitor session | User engagement depth metric |
| **TimeOnSite** | Numeric (minutes) | Average time spent on website per visit (in minutes) | Content engagement indicator |
| **SocialShares** | Integer | Number of times content was shared on social media | Content virality and organic reach metric |
| **EmailOpens** | Integer | Number of email opens (if email channel) | Email campaign engagement |
| **EmailClicks** | Integer | Number of clicks within email campaigns | Email campaign effectiveness |
| **PreviousPurchases** | Integer | Count of previous purchases by customer | Customer loyalty/history indicator |
| **LoyaltyPoints** | Numeric | Customer loyalty program points accumulated | Customer retention indicator |
| **AdvertisingPlatform** | String/Categorical | **Confidential** - Platform information not disclosed in dataset | Placeholder field for learning purposes |
| **AdvertisingTool** | String/Categorical | **Confidential** - Tool/feature information not disclosed in dataset | Placeholder field for learning purposes |
| **Conversion** | Binary (0/1) | Whether the campaign resulted in a conversion | **Target variable** for predictive modeling |


In [ ]:
# 4. DATA QUALITY ASSESSMENT

print("=" * 80)
print("DATA QUALITY ASSESSMENT")
print("=" * 80)

# A. Missing Values Analysis
print("\n--- A. Missing Values ---")
missing_counts = predict_conversion_df.isnull().sum()
missing_pct = (predict_conversion_df.isnull().sum() / len(predict_conversion_df)) * 100
missing_df = pd.DataFrame({
    'Column': missing_counts.index,
    'Missing_Count': missing_counts.values,
    'Percentage': missing_pct.values
})
missing_df = missing_df[missing_df['Missing_Count'] > 0]
if len(missing_df) == 0:
    print("✓ No missing values detected in any column")
else:
    print(missing_df.to_string(index=False))

# B. Duplicate Records Analysis
print("\n--- B. Duplicate Records ---")
duplicates = predict_conversion_df.duplicated().sum()
duplicate_pct = (duplicates / len(predict_conversion_df)) * 100
print(f"Duplicate rows: {duplicates} ({duplicate_pct:.2f}%)")

# C. Categorical Variables Analysis
print("\n--- C. Categorical Variables ---")
categorical_cols = predict_conversion_df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    unique_count = predict_conversion_df[col].nunique()
    print(f"\n{col}:")
    print(f"  - Unique values: {unique_count}")
    print(f"  - Value counts:\n{predict_conversion_df[col].value_counts()}")

# D. Numerical Variables - Range and Distribution
print("\n--- D. Numerical Variables - Ranges and Outliers ---")
numerical_cols = predict_conversion_df.select_dtypes(include=['int64', 'float64']).columns
for col in numerical_cols:
    data = predict_conversion_df[col]
    Q1 = data.quantile(0.25)
    Q3 = data.quantile(0.75)
    IQR = Q3 - Q1
    outliers = ((data < (Q1 - 1.5 * IQR)) | (data > (Q3 + 1.5 * IQR))).sum()
    print(f"\n{col}:")
    print(f"  - Min: {data.min():.2f}, Max: {data.max():.2f}")
    print(f"  - Mean: {data.mean():.2f}, Median: {data.median():.2f}, Std: {data.std():.2f}")
    print(f"  - Outliers (IQR method): {outliers} ({(outliers/len(data)*100):.2f}%)")

# E. Conversion Rate Balance
print("\n--- E. Target Variable Distribution ---")
if 'Conversion' in predict_conversion_df.columns:
    conversion_counts = predict_conversion_df['Conversion'].value_counts()
    conversion_pct = (predict_conversion_df['Conversion'].value_counts(normalize=True) * 100)
    print(f"Conversion Distribution:")
    for value in conversion_counts.index:
        print(f"  - {value}: {conversion_counts[value]} ({conversion_pct[value]:.2f}%)")


DATA QUALITY ASSESSMENT

--- A. Missing Values ---
✓ No missing values detected in any column

--- B. Duplicate Records ---
Duplicate rows: 0 (0.00%)

--- C. Categorical Variables ---

Gender:
  - Unique values: 2
  - Value counts:
Gender
Female    4839
Male      3161
Name: count, dtype: int64

CampaignChannel:
  - Unique values: 5
  - Value counts:
CampaignChannel
Referral        1719
PPC             1655
Email           1557
SEO             1550
Social Media    1519
Name: count, dtype: int64

CampaignType:
  - Unique values: 4
  - Value counts:
CampaignType
Conversion       2077
Awareness        1988
Consideration    1988
Retention        1947
Name: count, dtype: int64

AdvertisingPlatform:
  - Unique values: 1
  - Value counts:
AdvertisingPlatform
IsConfid    8000
Name: count, dtype: int64

AdvertisingTool:
  - Unique values: 1
  - Value counts:
AdvertisingTool
ToolConfid    8000
Name: count, dtype: int64

--- D. Numerical Variables - Ranges and Outliers ---

CustomerID:
  - Min: 80

## Data Quality Findings - Summary

### Strengths:
- ✓ **Complete dataset**: No missing values detected - dataset is clean by design
- ✓ **No duplicates**: All 8,000 customer records are unique
- ✓ **Balanced categorical variables**: Good representation across channels and campaign types (Email, Social Media, SEO, PPC, Referral)
- ✓ **Good sample size**: 8,000 records provide sufficient statistical power for predictive modeling
- ✓ **Well-structured for ML**: No outliers (IQR method), normalized distributions suitable for model training

### Limitations & Important Considerations:
- ⚠️ **Synthetic Data**: This is simulated data generated for educational purposes - does NOT represent real-world data
- ⚠️ **Confidential Fields**: AdvertisingPlatform and AdvertisingTool contain placeholder values ("IsConfid", "ToolConfid") as per dataset design
- ⚠️ **No temporal data**: Missing timestamp/date information - cannot perform time-series analysis
- ⚠️ **No geographic data**: No location information - cannot perform geographic segmentation
- ⚠️ **Class Imbalance**: Conversion variable is highly imbalanced (87.65% positive, 12.35% negative) - typical of synthetic balanced datasets
- ⚠️ **Not for real-world deployment**: This dataset is intended for learning and skill development, not for actual business decision-making
- ⚠️ **Educational context**: Results and patterns discovered should be validated against real business data before implementation


## 5. ETHICAL CONSIDERATIONS

### Important Context: Synthetic Data for Learning
This is a synthetically generated dataset for educational purposes. While ethical considerations may not apply directly to this synthetic data, they are **essential when applying the techniques learned here to real-world data**. Use this dataset to develop ethical data science practices.

### Data Privacy & Consent (for real-world applications)
- **Customer Data**: Real marketing datasets contain personally identifiable information (PII) including age, gender, income, and purchase history
- **Recommendation**: Ensure all data collection complies with GDPR, CCPA, and other privacy regulations
### Bias & Fairness (for real-world applications)
- **Demographic Targeting Risk**: Use of Age, Gender, and Income for targeting in real data could introduce discriminatory practices
- **Potential Harm**: Certain groups could be systematically excluded from campaigns or offered unfair terms
- **Mitigation Strategy**: Implement fairness constraints in modeling to prevent demographic discrimination
- **Learning Opportunity**: Use this synthetic data to practice building fairness audits and demographic parity checks before applying to real customer data
- **Transparency**: Clearly communicate targeting criteria to stakeholders and customers

### Marketing Ethics
- **Manipulation Concerns**: CTR and ConversionRate optimizations could lead to manipulative marketing tactics
- **User Experience**: Consider negative impacts of aggressive retargeting or frequency capping
- **Value Proposition**: Ensure campaigns deliver genuine value, not just maximize conversions
- **Honest Advertising**: Maintain transparency in advertising claims and avoid deceptive practices

### Data Security
- **Sensitive Information**: Income and previous purchase data are sensitive and require robust protection
- **Unauthorized Access**: Implement access controls and encryption for data security
- **Data Breach Risk**: Establish incident response procedures for potential breaches

### Model Usage & Decision-Making
- **Algorithmic Accountability**: If this data is used for predictive modeling, results should be audited for fairness
- **Explainability**: Marketing decisions should be explainable and challengeable by customers
- **Human Oversight**: Automated marketing decisions should have human review processes
- **Impact Assessment**: Evaluate potential negative impacts of marketing strategies on customer relationships

### Transparency & Accountability
- **Clear Communication**: Customers should understand why they're being targeted with specific campaigns
- **Opt-out Mechanisms**: Provide easy ways for customers to opt out of data-driven marketing
- **Data Usage Policy**: Maintain clear privacy policies about how customer data is used
- **Regulatory Compliance**: Stay updated with evolving data protection and marketing regulations
### Recommended Best Practices for Real-World Application
1. ✓ **Privacy-by-Design**: Implement privacy-by-design principles from the start of any real marketing system
2. ✓ **Fairness Audits**: Conduct regular audits for bias and fairness in targeting and modeling (practice with this synthetic data)
3. ✓ **Informed Consent**: Obtain explicit, informed consent from real customers before using their data
4. ✓ **Opt-out Mechanisms**: Provide transparent and easy opt-out mechanisms for data-driven marketing
5. ✓ **Impact Monitoring**: Monitor campaign impacts to avoid customer fatigue or negative sentiment
6. ✓ **Data Security**: Maintain robust data security and implement strict access controls
7. ✓ **Regulatory Compliance**: Regularly review compliance with GDPR, CCPA, and other privacy laws
8. ✓ **Customer-Centric**: Prioritize customer welfare and value delivery, not just conversion maximization
9. ✓ **Attribution**: When using this dataset or publishing results, properly cite Rabie El Kharoua under CC BY 4.0 license



7. ✓ Regularly review compliance with GDPR, CCPA, and other privacy laws
8. ✓ Consider customer welfare, not just conversion maximization

# Google Ads Analytics Dataset - Comprehensive Analysis

## 1. Data Sources and Collection Methods

**Dataset Name:** GoogleAds_DataAnalytics_Sales_Uncleaned.csv

**Source & Nature:** This is a **simulated Google Ads campaign dataset** promoting data analytics courses and services. The dataset **intentionally mimics real-world Google Ads export data** with all its messiness and quality issues.

**Purpose & Design:** 
- Designed to simulate what real digital marketers encounter when exporting campaign data
- Intentionally left **raw and uncleaned** to teach data cleaning and preprocessing techniques
- Perfect for practicing: data cleaning, EDA, marketing analytics, and dashboard creation
- Campaign Focus: Promotion of **data analytics courses/services**

**Data Simulation:** Represents ad performance metrics for a data analytics course campaign across:
- Multiple campaign variations (with intentional typos)
- Geographic locations (with spelling/case variations)
- Device types (with mixed casing)
- Keywords (with spelling errors)
- Time period: November 2024 campaign data

**Data Characteristics:**
- Ad-level granularity (2,600 unique ads)
- Advertising performance metrics (clicks, impressions, cost)
- Business outcome metrics (leads, conversions, revenue)
- Temporal dimension (Ad_Date in inconsistent formats)
- Geographic and device segmentation

- Keyword-level performance data**Intentional Quality Issues:** The dataset contains real-world messiness including typos, inconsistent formats, missing values, and duplicate rows - all designed for learning data cleaning techniques

- Mixed currency formats (₹ or $ symbols)

**License:** CC0 - Public Domain (free to use without attribution)

In [9]:
# 2. DATA DESCRIPTION: Summary Statistics, Data Types, and Volume

print("=" * 80)
print("GOOGLE ADS DATASET OVERVIEW")
print("=" * 80)
print(f"\nDataset Dimensions:")
print(f"  - Rows (Samples/Ads): {google_ads_df.shape[0]:,}")
print(f"  - Columns (Features): {google_ads_df.shape[1]}")
print(f"  - Total Data Points: {google_ads_df.shape[0] * google_ads_df.shape[1]:,}")

print("\n" + "=" * 80)
print("COLUMN NAMES AND DATA TYPES")
print("=" * 80)
for col in google_ads_df.columns:
    print(f"  - {col}: {google_ads_df[col].dtype}")

print("\n" + "=" * 80)
print("DATA TYPES SUMMARY")
print("=" * 80)
print(google_ads_df.dtypes)

print("\n" + "=" * 80)
print("SUMMARY STATISTICS (Numerical Columns)")
print("=" * 80)
print(google_ads_df.describe().T)

print("\n" + "=" * 80)
print("FIRST FEW RECORDS")
print("=" * 80)
print(google_ads_df.head())


GOOGLE ADS DATASET OVERVIEW

Dataset Dimensions:
  - Rows (Samples/Ads): 2,600
  - Columns (Features): 13
  - Total Data Points: 33,800

COLUMN NAMES AND DATA TYPES
  - Ad_ID: object
  - Campaign_Name: object
  - Clicks: float64
  - Impressions: float64
  - Cost: object
  - Leads: float64
  - Conversions: float64
  - Conversion Rate: float64
  - Sale_Amount: object
  - Ad_Date: object
  - Location: object
  - Device: object
  - Keyword: object

DATA TYPES SUMMARY
Ad_ID               object
Campaign_Name       object
Clicks             float64
Impressions        float64
Cost                object
Leads              float64
Conversions        float64
Conversion Rate    float64
Sale_Amount         object
Ad_Date             object
Location            object
Device              object
Keyword             object
dtype: object

SUMMARY STATISTICS (Numerical Columns)
                  count         mean         std       min       25%  \
Clicks           2488.0   138.956994   34.619376    80.

## 3. DATA DICTIONARY - Google Ads Dataset

| Column Name | Data Type | Description | Relevance to Analysis |
|---|---|---|---|
| **Ad_ID** | Integer/String | Unique identifier for each advertisement | Key for tracking individual ads and joins with other tables |
| **Campaign_Name** | String/Categorical | Name/identifier of the advertising campaign | Campaign-level aggregation and performance comparison |
| **Clicks** | Integer | Number of times users clicked on the ad | **Key Metric** - Traffic generation indicator |
| **Impressions** | Integer | Number of times the ad was displayed to users | **Key Metric** - Ad reach and visibility measure |
| **Cost** | String (with $ or ₹) | Total cost spent on ad (₹ or $ format, **missing values present**) | **Key Metric** - Budget analysis and ROI (requires conversion/cleaning) |
| **Leads** | Numeric | Number of leads generated from the ad | **Business Metric** - Sales funnel indicator |
| **Conversions** | Numeric | Number of actual conversions/signups from the ad | **Key Metric** - Campaign effectiveness measure |
| **Conversion Rate** | Numeric | Conversions ÷ Clicks, **missing in ~24% of records** | **Key KPI** - Conversion efficiency metric |
| **Sale_Amount** | String (with $ or ₹) | Revenue generated from conversions (**missing values present**) | **Key Metric** - Revenue impact and ROI analysis |
| **Ad_Date** | String (mixed formats) | Date of ad activity (**inconsistent formats: YYYY/MM/DD, DD-MM-YYYY, DD-MM-YY**) | **Temporal Dimension** - Requires date parsing and standardization |
| **Location** | String/Categorical | City where ad served (**includes spelling/case variations: Hyderabad, hyderabad, Hyderbad, hydrebad**) | Geographic segmentation (requires standardization) |
| **Device** | String/Categorical | Device type (**mixed casing: Mobile, mobile, MOBILE, Desktop, desktop, DESKTOP, Tablet, tablet, TABLET**) | Device-specific performance analysis (requires standardization) |
| **Keyword** | String/Categorical | Search/targeting keyword (**with typos: analitics, anaytics instead of analytics**) | Keyword performance analysis (requires keyword cleaning) |


In [10]:
# 4. DATA QUALITY ASSESSMENT - Google Ads Dataset

print("=" * 80)
print("DATA QUALITY ASSESSMENT - GOOGLE ADS")
print("=" * 80)

# A. Missing Values Analysis
print("\n--- A. Missing Values ---")
missing_counts = google_ads_df.isnull().sum()
missing_pct = (google_ads_df.isnull().sum() / len(google_ads_df)) * 100
missing_df = pd.DataFrame({
    'Column': missing_counts.index,
    'Missing_Count': missing_counts.values,
    'Percentage': missing_pct.values
})
missing_df = missing_df[missing_df['Missing_Count'] > 0].sort_values('Missing_Count', ascending=False)
if len(missing_df) == 0:
    print("✓ No missing values detected in any column")
else:
    print(f"⚠️ Missing values detected:\n{missing_df.to_string(index=False)}")

# B. Duplicate Records Analysis
print("\n--- B. Duplicate Records ---")
duplicates = google_ads_df.duplicated().sum()
duplicate_pct = (duplicates / len(google_ads_df)) * 100
print(f"Duplicate rows: {duplicates} ({duplicate_pct:.2f}%)")
if duplicates > 0:
    print(f"⚠️ Found {duplicates} duplicate records that may need investigation")

# C. Categorical Variables Analysis
print("\n--- C. Categorical Variables ---")
categorical_cols = google_ads_df.select_dtypes(include=['object']).columns
if len(categorical_cols) > 0:
    for col in categorical_cols:
        unique_count = google_ads_df[col].nunique()
        print(f"\n{col}:")
        print(f"  - Unique values: {unique_count}")
        if unique_count <= 20:
            print(f"  - Value counts:\n{google_ads_df[col].value_counts()}")
        else:
            print(f"  - Top 10 values:\n{google_ads_df[col].value_counts().head(10)}")

# D. Numerical Variables - Range and Distribution
print("\n--- D. Numerical Variables - Ranges and Outliers ---")
numerical_cols = google_ads_df.select_dtypes(include=['int64', 'float64']).columns
for col in numerical_cols:
    data = google_ads_df[col]
    Q1 = data.quantile(0.25)
    Q3 = data.quantile(0.75)
    IQR = Q3 - Q1
    outliers = ((data < (Q1 - 1.5 * IQR)) | (data > (Q3 + 1.5 * IQR))).sum()
    print(f"\n{col}:")
    print(f"  - Min: {data.min():.2f}, Max: {data.max():.2f}")
    print(f"  - Mean: {data.mean():.2f}, Median: {data.median():.2f}, Std: {data.std():.2f}")
    if data.min() < 0:
        print(f"  - ⚠️ WARNING: Negative values detected (Min: {data.min()})")
    if outliers > 0:
        print(f"  - Outliers (IQR method): {outliers} ({(outliers/len(data)*100):.2f}%)")

# E. Date Format Analysis
print("\n--- E. Temporal Data ---")
if 'Ad_Date' in google_ads_df.columns:
    print(f"Ad_Date column dtype: {google_ads_df['Ad_Date'].dtype}")
    print(f"Sample dates: {google_ads_df['Ad_Date'].head()}")
    try:
        dates = pd.to_datetime(google_ads_df['Ad_Date'])
        print(f"Date range: {dates.min()} to {dates.max()}")
        print(f"Time span: {(dates.max() - dates.min()).days} days")
    except:
        print("⚠️ Unable to parse dates - may need date format conversion")

# F. Logic Consistency Checks
print("\n--- F. Data Logic Consistency ---")
if 'Clicks' in google_ads_df.columns and 'Impressions' in google_ads_df.columns:
    invalid_ctr = (google_ads_df['Clicks'] > google_ads_df['Impressions']).sum()
    if invalid_ctr > 0:
        print(f"⚠️ {invalid_ctr} records have more Clicks than Impressions (logical error)")
    else:
        print("✓ Clicks <= Impressions (logically consistent)")

if 'Conversions' in google_ads_df.columns and 'Clicks' in google_ads_df.columns:
    invalid_conv = (google_ads_df['Conversions'] > google_ads_df['Clicks']).sum()
    if invalid_conv > 0:
        print(f"⚠️ {invalid_conv} records have more Conversions than Clicks (logical error)")
    else:
        print("✓ Conversions <= Clicks (logically consistent)")


DATA QUALITY ASSESSMENT - GOOGLE ADS

--- A. Missing Values ---
⚠️ Missing values detected:
         Column  Missing_Count  Percentage
Conversion Rate            626   24.076923
    Sale_Amount            139    5.346154
         Clicks            112    4.307692
           Cost             97    3.730769
    Conversions             74    2.846154
    Impressions             54    2.076923
          Leads             48    1.846154

--- B. Duplicate Records ---
Duplicate rows: 0 (0.00%)

--- C. Categorical Variables ---

Ad_ID:
  - Unique values: 2600
  - Top 10 values:
Ad_ID
A1000    1
A1001    1
A1002    1
A1003    1
A1004    1
A1005    1
A1006    1
A1007    1
A1008    1
A1009    1
Name: count, dtype: int64

Campaign_Name:
  - Unique values: 4
  - Value counts:
Campaign_Name
Data Analytcis Course    680
Data Analytics Corse     647
DataAnalyticsCourse      637
Data Anlytics Corse      636
Name: count, dtype: int64

Cost:
  - Unique values: 2106
  - Top 10 values:
Cost
$246.3     4
$2

## Data Quality Findings - Summary

### Dataset Purpose: Learning Data Cleaning Skills

**Dataset Status:** "Uncleaned" simulation - **intentionally contains real-world data messiness** for educational purposes

**Sample Size:** 2,600 ad records for a data analytics course promotion campaign

### Intentional Data Quality Issues (Learning Opportunities):

#### Missing Values (practice data imputation/removal):
- ⚠️ **Conversion Rate**: Missing in ~24% of records (requires calculation)
- ⚠️ **Sale_Amount**: Missing in ~5.3% of records
- ⚠️ **Clicks, Cost, Conversions, Impressions, Leads**: Missing in 2-4% each

#### Format & Encoding Issues (practice standardization):
- ⚠️ **Ad_Date**: 3+ inconsistent formats (YYYY/MM/DD, DD-MM-YYYY, DD-MM-YY) - requires date parsing
- ⚠️ **Cost & Sale_Amount**: Stored as strings with currency symbols ($ or ₹) - requires type conversion
- ⚠️ **Device**: Mixed casing (MOBILE, mobile, Mobile, DESKTOP, Desktop, desktop, TABLET, tablet, Tablet)
- ⚠️ **Location**: Spelling/case variations (HYDERABAD, hyderabad, Hyderbad, hydrebad)
- ⚠️ **Campaign_Name**: Typos and variations (Analytics, Anlytics, Analytcis, Analtyics, Corse, Course)
- ⚠️ **Keyword**: Spelling errors (analitics, anaytics, anadlytic instead of analytics)

#### Data Consistency Issues (practice validation):
- ⚠️ **Duplicate Rows**: May contain duplicates - practice duplicate detection and removal
- ⚠️ **Logical Consistency**: Verify Clicks ≤ Impressions, Conversions ≤ Clicks
- ⚠️ **Value Ranges**: Check for negative/zero values where illogical

### Learning Objectives:
1. ✓ Practice data cleaning with real-world messy data
2. ✓ Learn to standardize categorical variables
3. ✓ Handle missing values appropriately
4. ✓ Parse and convert inconsistent date formats
5. ✓ Convert currency strings to numeric values
6. ✓ Detect and handle duplicates
7. ✓ Validate logical constraints in data
8. ✓ Prepare data for exploratory analysis and modeling


## 5. ETHICAL CONSIDERATIONS - Learning Data Cleaning

### Educational Purpose & Simulation Context
**This is a simulated dataset for learning**, not real user data. The data quality issues are intentional and designed for educational practice. When applying these techniques to real data:

### Data Quality as an Ethical Issue
- **Garbage In, Garbage Out**: Poor quality data can lead to incorrect business insights
- **Bias Introduction**: Uncleaned data with patterns can amplify existing biases in analysis
- **Decision Impact**: Companies making decisions on dirty data risk harming customers or operations
- **Professional Responsibility**: Data professionals have ethical obligation to clean/validate data before analysis

### Learning to Handle Real-World Data Issues

#### Typos & Standardization (Practice Here):
- Learning to standardize variations teaches precision in analysis
- Real-world example: Device categorization errors can skew device-level insights
- Practice: Create standardization rules that can be applied to production data

#### Missing Values (Practice Here):
- ~24% missing Conversion Rate reflects real business scenarios
- Decisions on handling missing data have downstream impacts
- Practice: Document assumptions about missing values clearly for stakeholders

#### Date Format Issues (Practice Here):
- Date parsing errors can invalidate time-series analysis
- Different regions use different date conventions - real teams must handle this
- Practice: Create robust date parsing that handles multiple formats

#### Currency Conversion (Practice Here):
- Real teams work with multiple currencies (₹ vs $)
- Failure to handle currency properly leads to massive financial errors
- Practice: Implement proper currency conversion logic with clear documentation

### Privacy & Compliance (Learning Perspective)
- This simulated data removes sensitive PII but represents ad campaigns
- Real Google Ads data requires compliance with GDPR, CCPA, and advertising regulations
- Learning how to clean data helps create processes that maintain privacy

### Best Practices for Professional Data Cleaning
1. ✓ **Document Everything**: Record all cleaning decisions and transformations
2. ✓ **Create Audit Trail**: Keep before/after versions to verify changes
3. ✓ **Validate Logic**: Ensure cleaning rules don't introduce new errors
4. ✓ **Test Assumptions**: Verify any decisions about missing/invalid data
5. ✓ **Standardize Consistently**: Apply same rules across entire dataset
6. ✓ **Handle Duplicates Carefully**: Understand why duplicates exist before removing
7. ✓ **Preserve Intent**: Keep original data when possible, create cleaned versions
8. ✓ **Communicate Issues**: Report data quality findings to stakeholders
9. ✓ **Reusable Processes**: Create scripts/functions that can be applied to future data
10. ✓ **Ethical Analysis**: Use cleaned data responsibly to avoid perpetuating biases


# Marketing and Product Performance Dataset - Comprehensive Analysis

## 1. Data Sources and Collection Methods

**Dataset Name:** marketing_and_product_performance.csv

**Source & Nature:** This is a **synthetically generated dataset** created by Imran Ali Shah for educational and research purposes. It contains artificially generated data designed to protect privacy while maintaining statistical integrity and realistic marketing scenarios.

**Purpose & Design:**
- Designed to assist researchers, data analysts, and marketers in evaluating campaign efficiency
- Provides a rich basis for analysis and experimentation in marketing contexts
- Perfect for learning data exploration, cleaning, and analysis techniques
- Simulates real-world marketing scenarios without exposing sensitive customer data

**Data Coverage:** 10,000 rows of synthetically generated campaign-product performance combinations

**Data Dimensions:**
- **Campaign Performance**: Budgets, clicks, conversions, ROI, and revenue generated
- **Product Details**: Units sold, discount levels, and bundle pricing
- **Customer Insights**: Subscription tiers, lengths, satisfaction ratings, and purchasing behavior
- **Promotional Context**: Flash sales, discount levels, and common keyword themes

**Intended Applications:**
- Marketing Analysis: Optimize campaigns and understand product success metrics
- Predictive Modeling: Train models for conversion rate prediction, revenue forecasting, and customer segmentation
- Educational Use: Practice data exploration, cleaning, and analysis in a realistic marketing context

**Data Quality Notes:** All data points are artificially generated to maintain statistical integrity while protecting privacy. Some logical inconsistencies may be present to simulate real-world data quality challenges.

**License:** CC BY 4.0 (Attribution 4.0 International) - Attribution required when using

**Author:** Imran Ali Shah

In [11]:
# 2. DATA DESCRIPTION: Summary Statistics, Data Types, and Volume

print("=" * 80)
print("MARKETING & PRODUCT PERFORMANCE DATASET OVERVIEW")
print("=" * 80)
print(f"\nDataset Dimensions:")
print(f"  - Rows (Campaign-Product Pairs): {marketing_performance_df.shape[0]:,}")
print(f"  - Columns (Features): {marketing_performance_df.shape[1]}")
print(f"  - Total Data Points: {marketing_performance_df.shape[0] * marketing_performance_df.shape[1]:,}")

print("\n" + "=" * 80)
print("COLUMN NAMES AND DATA TYPES")
print("=" * 80)
for col in marketing_performance_df.columns:
    print(f"  - {col}: {marketing_performance_df[col].dtype}")

print("\n" + "=" * 80)
print("DATA TYPES SUMMARY")
print("=" * 80)
print(marketing_performance_df.dtypes)

print("\n" + "=" * 80)
print("SUMMARY STATISTICS (Numerical Columns)")
print("=" * 80)
print(marketing_performance_df.describe().T)

print("\n" + "=" * 80)
print("FIRST FEW RECORDS")
print("=" * 80)
print(marketing_performance_df.head())


MARKETING & PRODUCT PERFORMANCE DATASET OVERVIEW

Dataset Dimensions:
  - Rows (Campaign-Product Pairs): 10,000
  - Columns (Features): 17
  - Total Data Points: 170,000

COLUMN NAMES AND DATA TYPES
  - Campaign_ID: object
  - Product_ID: object
  - Budget: float64
  - Clicks: int64
  - Conversions: int64
  - Revenue_Generated: float64
  - ROI: float64
  - Customer_ID: object
  - Subscription_Tier: object
  - Subscription_Length: int64
  - Flash_Sale_ID: object
  - Discount_Level: int64
  - Units_Sold: int64
  - Bundle_ID: object
  - Bundle_Price: float64
  - Customer_Satisfaction_Post_Refund: int64
  - Common_Keywords: object

DATA TYPES SUMMARY
Campaign_ID                           object
Product_ID                            object
Budget                               float64
Clicks                                 int64
Conversions                            int64
Revenue_Generated                    float64
ROI                                  float64
Customer_ID                   

## 3. DATA DICTIONARY - Marketing and Product Performance Dataset

| Column Name | Data Type | Description | Relevance to Analysis |
|---|---|---|---|
| **Campaign_ID** | Integer | Unique identifier for each marketing campaign | Key for campaign-level aggregation and tracking |
| **Product_ID** | Integer | Unique identifier for the product being promoted | Product-level performance analysis and segmentation |
| **Budget** | Numeric (Currency) | Total budget allocated to the campaign | **Key Metric** - Campaign investment and ROI calculations |
| **Clicks** | Integer | Number of clicks received on the campaign | Traffic generation and engagement metric |
| **Conversions** | Integer | Number of conversions from the campaign | Campaign effectiveness metric |
| **Revenue_Generated** | Numeric (Currency) | Total revenue produced by the campaign | **Key Metric** - Business outcome and ROI analysis |
| **ROI** | Numeric (%) | Return on Investment = (Revenue - Budget) / Budget | **Key KPI** - Campaign profitability metric |
| **Customer_ID** | Integer | Identifier of customer associated with campaign | Customer segmentation and analysis |
| **Subscription_Tier** | String/Categorical | Service/subscription level (Basic, Pro, Premium, etc.) | Customer value segmentation |
| **Subscription_Length** | Integer | Duration of subscription in months | Customer lifetime value and retention indicator |
| **Flash_Sale_ID** | Integer | Identifier for promotional flash sale (if applicable) | Promotion strategy and impact analysis |
| **Discount_Level** | Numeric (%) | Percentage discount applied (if any) | Promotion effectiveness and margin impact |
| **Units_Sold** | Integer | Number of units sold through the campaign | Sales volume metric |
| **Bundle_ID** | Integer | Identifier for product bundle (if applicable) | Bundle strategy and cross-sell analysis |
| **Bundle_Price** | Numeric (Currency) | Price of bundled products | Bundle pricing and revenue impact |
| **Customer_Satisfaction_Post_Refund** | Numeric (0-100 or score) | Customer satisfaction score after refund/resolution | **Quality Metric** - Customer experience indicator |
| **Common_Keywords** | String/Categorical | Keywords used in the campaign | SEO/SEM strategy and keyword effectiveness |


In [12]:
# 4. DATA QUALITY ASSESSMENT - Marketing and Product Performance

print("=" * 80)
print("DATA QUALITY ASSESSMENT - MARKETING & PRODUCT PERFORMANCE")
print("=" * 80)

# A. Missing Values Analysis
print("\n--- A. Missing Values ---")
missing_counts = marketing_performance_df.isnull().sum()
missing_pct = (marketing_performance_df.isnull().sum() / len(marketing_performance_df)) * 100
missing_df = pd.DataFrame({
    'Column': missing_counts.index,
    'Missing_Count': missing_counts.values,
    'Percentage': missing_pct.values
})
missing_df = missing_df[missing_df['Missing_Count'] > 0].sort_values('Missing_Count', ascending=False)
if len(missing_df) == 0:
    print("✓ No missing values detected in any column")
else:
    print(f"⚠️ Missing values detected:\n{missing_df.to_string(index=False)}")

# B. Duplicate Records Analysis
print("\n--- B. Duplicate Records ---")
duplicates = marketing_performance_df.duplicated().sum()
duplicate_pct = (duplicates / len(marketing_performance_df)) * 100
print(f"Duplicate rows: {duplicates} ({duplicate_pct:.2f}%)")
if duplicates > 0:
    print(f"⚠️ Found {duplicates} duplicate records")

# C. Categorical Variables Analysis
print("\n--- C. Categorical Variables ---")
categorical_cols = marketing_performance_df.select_dtypes(include=['object']).columns
if len(categorical_cols) > 0:
    for col in categorical_cols:
        unique_count = marketing_performance_df[col].nunique()
        print(f"\n{col}:")
        print(f"  - Unique values: {unique_count}")
        if unique_count <= 20:
            print(f"  - Value counts:\n{marketing_performance_df[col].value_counts()}")
        else:
            print(f"  - Top 10 values:\n{marketing_performance_df[col].value_counts().head(10)}")

# D. Numerical Variables - Range and Distribution
print("\n--- D. Numerical Variables - Ranges and Outliers ---")
numerical_cols = marketing_performance_df.select_dtypes(include=['int64', 'float64']).columns
for col in numerical_cols:
    data = marketing_performance_df[col]
    Q1 = data.quantile(0.25)
    Q3 = data.quantile(0.75)
    IQR = Q3 - Q1
    outliers = ((data < (Q1 - 1.5 * IQR)) | (data > (Q3 + 1.5 * IQR))).sum()
    print(f"\n{col}:")
    print(f"  - Min: {data.min():.2f}, Max: {data.max():.2f}")
    print(f"  - Mean: {data.mean():.2f}, Median: {data.median():.2f}, Std: {data.std():.2f}")
    if data.min() < 0:
        print(f"  - ⚠️ WARNING: Negative values detected (Min: {data.min()})")
    if outliers > 0:
        print(f"  - Outliers (IQR method): {outliers} ({(outliers/len(data)*100):.2f}%)")

# E. Business Metric Validation
print("\n--- E. Business Metric Validation ---")
if 'Revenue_Generated' in marketing_performance_df.columns and 'Budget' in marketing_performance_df.columns:
    negative_revenue = (marketing_performance_df['Revenue_Generated'] < 0).sum()
    if negative_revenue > 0:
        print(f"⚠️ {negative_revenue} records with negative revenue")
    else:
        print("✓ All revenue values are non-negative")

if 'ROI' in marketing_performance_df.columns:
    invalid_roi = marketing_performance_df['ROI'].isnull().sum()
    print(f"ROI missing values: {invalid_roi}")
    roi_data = marketing_performance_df['ROI'].dropna()
    if len(roi_data) > 0:
        print(f"  - Min ROI: {roi_data.min():.2f}%, Max ROI: {roi_data.max():.2f}%")
        print(f"  - Mean ROI: {roi_data.mean():.2f}%, Median ROI: {roi_data.median():.2f}%")

# F. Campaign Metrics Consistency
print("\n--- F. Campaign Metrics Consistency ---")
if 'Conversions' in marketing_performance_df.columns and 'Clicks' in marketing_performance_df.columns:
    invalid_conv = (marketing_performance_df['Conversions'] > marketing_performance_df['Clicks']).sum()
    if invalid_conv > 0:
        print(f"⚠️ {invalid_conv} records have more Conversions than Clicks (logical error)")
    else:
        print("✓ Conversions <= Clicks (logically consistent)")

if 'Units_Sold' in marketing_performance_df.columns:
    zero_sales = (marketing_performance_df['Units_Sold'] == 0).sum()
    print(f"Zero units sold: {zero_sales} records ({(zero_sales/len(marketing_performance_df)*100):.2f}%)")


DATA QUALITY ASSESSMENT - MARKETING & PRODUCT PERFORMANCE

--- A. Missing Values ---
✓ No missing values detected in any column

--- B. Duplicate Records ---
Duplicate rows: 0 (0.00%)

--- C. Categorical Variables ---

Campaign_ID:
  - Unique values: 10000
  - Top 10 values:
Campaign_ID
CMP_RLSDVN    1
CMP_JHHUE9    1
CMP_6SBOWN    1
CMP_Q31QCU    1
CMP_AY0UTJ    1
CMP_5F4E9W    1
CMP_WDWHH6    1
CMP_JECPN0    1
CMP_WWJQBU    1
CMP_G2GXC0    1
Name: count, dtype: int64

Product_ID:
  - Unique values: 10000
  - Top 10 values:
Product_ID
PROD_HBJFA3    1
PROD_OE8YNJ    1
PROD_4V8A08    1
PROD_A1Q6ZB    1
PROD_F57N66    1
PROD_2CF719    1
PROD_QN8SL7    1
PROD_FN0KK1    1
PROD_12IBBU    1
PROD_HBPOK5    1
Name: count, dtype: int64

Customer_ID:
  - Unique values: 10000
  - Top 10 values:
Customer_ID
CUST_1K7G39    1
CUST_0DWS6F    1
CUST_BR2GST    1
CUST_6TBY6K    1
CUST_XASI45    1
CUST_GH04I5    1
CUST_TRQ4H8    1
CUST_XOGUSD    1
CUST_5RON6U    1
CUST_VPSNHS    1
Name: count, dtype: in

## Data Quality Findings - Summary

### Dataset Characteristics:

**Dataset Status:** Synthetically generated educational data designed for learning and analysis

**Sample Size:** 10,000 campaign-product pairs with 17 performance and business metrics

**Data Completeness:** All columns have complete data (0% missing) - typical of synthetic datasets

### Data Quality Strengths:

**✓ Comprehensive Coverage:**
- Large sample size (10,000 records) suitable for robust analysis
- Multi-dimensional data (campaign, product, customer, promotion dimensions)
- Complete business metrics (Budget, Revenue, ROI) for financial analysis
- Customer satisfaction data for quality assessment (1-4 scale)
- Promotional strategy data (flash sales, bundles, discounts) for segmentation

**✓ Synthetic Data Advantages:**
- No missing values (100% complete data)
- No duplicate records
- Balanced categorical distributions (Subscription Tiers: ~33% each)
- Privacy-protected (no real customer data)
- Statistical integrity maintained for educational purposes

### Critical Data Quality Issues (Educational Value):

**⚠️ Logical Inconsistencies Identified:**
- **959 records (9.59%) have more Conversions than Clicks** - logically impossible in real marketing data
- This violation of business logic is useful for identifying data validation errors
- Demonstrates importance of validating assumptions before analysis
- Provides practical experience detecting and handling illogical data combinations

**Purpose of These Issues:**
- Intentionally included to simulate real-world data quality challenges
- Helps users develop data validation and cleaning skills
- Teaches identification of business logic violations
- Prepares analysts for handling messy real-world data

### Data Quality Observations:

**Key Findings:**
- **Subscription Tiers**: Balanced distribution (Basic: 34.2%, Standard: 33%, Premium: 32.8%)
- **Satisfaction Scale**: 1-4 scale with mean satisfaction of 2.5
- **ROI Distribution**: 0.5% to 5.0% with mean ROI of 2.76%
- **Budget Range**: $500-$50,000 (mean: $25,264)
- **Revenue Generated**: $1,002-$100,000 (mean: $50,039)
- **All IDs appear unique**: Campaign_ID, Product_ID, Customer_ID all have 10,000 unique values

### Validation Checks Needed:

1. **Logical Constraint Violations**: Investigate and decide handling strategy for Conversions > Clicks cases
2. **ROI Calculation**: Verify ROI = (Revenue - Budget) / Budget across all records
3. **Business Logic**: Validate Conversions ≤ Clicks, Units_Sold alignment
4. **Satisfaction Scale**: Confirm 1-4 scale interpretation for analysis
5. **Null Handling**: Verify Flash_Sale_ID, Bundle_ID, Discount_Level are truly complete or strategically null
6. **Outlier Analysis**: High ROI values are legitimate performance, not errors

### Educational Takeaways:

This dataset effectively teaches:
- **Data validation techniques** for identifying illogical combinations
- **Business constraint checking** for marketing metrics
- **Handling inconsistencies** in real-world data scenarios
- **Quality assessment methods** before analysis begins
- **Risk of using unvalidated data** in business decisions


## 5. ETHICAL CONSIDERATIONS - Synthetic Marketing Data for Learning

### Educational & Synthetic Data Context
**This is synthetically generated data designed for educational purposes.** The data is artificially generated to protect privacy while maintaining realistic marketing scenarios. These ethical considerations apply when using similar techniques with **real customer data**.

### Learning Value of This Synthetic Dataset
- Allows exploration of marketing ethics without involving real customers
- Enables practice with data quality issues (logical inconsistencies) for skill building
- Provides risk-free environment to learn marketing analytics techniques
- Helps develop intuition for data validation and quality assessment

### When Applying to Real Customer Data: Campaign & Product Targeting Ethics

#### Subscription Tier Discrimination (Real Data Concern):
- **Risk**: Different promotional offers by subscription tier could create unfair targeting
- **Concern**: Premium customers may receive different value than basic customers
- **Fairness**: Ensure pricing and promotions offer fair value at each tier
- **Transparency**: Disclose why different offers are shown to different tiers
- **Learning Opportunity**: Use this dataset to practice fair pricing analysis

#### Flash Sales & Discount Strategy (Real Data Concern):
- **Manipulation Risk**: Discounts may target vulnerable segments (e.g., price-sensitive customers)
- **Fairness Issue**: Artificial scarcity created by flash sales
- **Refund Fairness**: Satisfaction tracking shouldn't penalize refund requests
- **Price Discrimination**: Ensure discounts don't disadvantage certain customer groups
- **Learning Opportunity**: Analyze promotional effectiveness equitably across segments

#### Bundle Strategy Ethics (Real Data Concern):
- **Transparency**: Hidden costs in complex bundles can confuse customers
- **Forced Purchasing**: Bundling may pressure customers to buy unwanted items
- **Consumer Confusion**: Complex bundles may be intentionally confusing
- **Competition Impact**: Bundling strategies might disadvantage competitors
- **Learning Opportunity**: Practice transparent bundle analysis and pricing

### Customer Data Privacy (Real Data Concern):
- **Customer Tracking**: Real data contains Customer_ID linking to behavior patterns
- **Consent & Transparency**: Customers should know how their data is tracked
- **Privacy Regulations**: GDPR, CCPA compliance required for real customer data
- **Satisfaction Data**: Post-refund satisfaction tracking could discourage legitimate refunds
- **Learning Opportunity**: This synthetic dataset teaches analysis without privacy risks

### ROI Optimization vs. Customer Welfare (Real Data Concern):
- **Revenue Focus**: Optimizing ROI might sacrifice customer retention and loyalty
- **Quality Trade-offs**: Profit maximization may reduce product/service quality
- **Churn Risk**: Aggressive optimization could harm customer relationships
- **Sustainable Balance**: Balance financial performance with customer welfare
- **Learning Opportunity**: Analyze ROI alongside customer satisfaction metrics

### Data-Driven Decision Making (Real Data Concern):
- **Algorithmic Bias**: Optimization algorithms might perpetuate biases
- **Informed Teams**: Marketing professionals should understand data limitations
- **Outlier Investigation**: Unusual metrics should be understood, not just optimized
- **Customer Impact**: Strategies should balance profit with customer benefit
- **Learning Opportunity**: Practice identifying and understanding data anomalies

### Best Practices for Ethical Marketing Analytics (Learning Outcomes)

1. ✓ **Data Validation First**: Always validate data quality before analysis (as shown by Conversions > Clicks inconsistency)
2. ✓ **Transparency**: Clearly communicate pricing, promotions, and targeting criteria to customers
3. ✓ **Fairness Audits**: Regular audits for disparate impact across customer segments
4. ✓ **Customer Choice**: Ensure customers can opt out of targeted promotions
5. ✓ **Quality Priority**: Prioritize product/service quality over short-term ROI
6. ✓ **Refund Respect**: Treat refund requests fairly, don't punish through satisfaction tracking
7. ✓ **Honest Marketing**: Use keywords and messaging that genuinely reflect product value
8. ✓ **Tier Fairness**: Ensure subscription tiers offer fair value at each price level
9. ✓ **Metric Understanding**: Understand limitations and biases in your data before using for decisions
10. ✓ **Sustainable Growth**: Balance profit optimization with customer retention and satisfaction
11. ✓ **Attribution**: Credit the dataset author (Imran Ali Shah) and cite CC BY 4.0 license when publishing
12. ✓ **Educational Purpose**: Remember this is synthetic data for learning - validate techniques with real data before deployment
